### PR TITLE
JENKINS-48583: Improve logic that handles suite aggregation

### PR DESCRIPTION
**Pull request by khneal**
Original intended use case was for complete duplicate suite definitions due to aggregation of multiple suites into a single file by ant's JUnitReport task. I added a test to ensure that scenario is still supported (who still uses ant? :D ).

Modern tools run tests in parallel, and a single suite may have test cases written across several files. The updated logic will only ignore a suite if it has an identical name, timestamp, duration, and number of test cases as the existing. If both suites being compared only have a single test case, then the full name of test case will also be compared.

Also, any ignored test suites are now logged as a warning. Many productive hours have been lost investigating this issue, so hopefully that will minimize future hours lost if another use case is discovered.

### CODE REVIEW

**Comment by khneal** [comment]
Edited commit messages to add square brackets around JIRA issue ID

**Comment by khneal** [comment]
FYI - it appears #20 was opened years ago to fix this same issue.

File: src/main/java/hudson/tasks/junit/TestResult.java [hunk 1]
```diff
@@ -286,16 +288,30 @@ private void add(SuiteResult sr) {
...
-                    return;
+                if(strictEq(s.getTimestamp(),sr.getTimestamp()) &&
+                        s.getDuration() == sr.getDuration() &&
+                        s.getCases().size() == sr.getCases().size()) {
+                    
+                    // Rare edge case: testing framework writes one file per test case, but all same test suite...
+                    // *and* duration is not included, or so fast that first two files have 0.000 or 0.001.
+                    // When the second file is read, number of test cases will match... so compare the name.
+                    if (s.getCases().size() != 1 ||
+                            s.getCases().get(0).getFullName().equals(sr.getCases().get(0).getFullName())) {
```
Comment (jonlil) [inline src/main/java/hudson/tasks/junit/TestResult.java]: I'm no java developer but this seems to be very complex. Perhaps extract this logic to a private method and give it a self explanatory method name? That would also make the logic easier to test.

Regard the implementation I'm not sure adding comparisons for duration and filename will make this problem go away. I think there is two paths to take here.

1. calculate checksums for the report files and compare them
2. disable this logic with a configuration parameter like

**Jenkinsfile**
```groovy
junit 'tmp/junit.xml', disable_duplicate_report_check: true
```

I would like to see option 1 since that actually solves this bug. 
This implementation just makes it even harder to understand and debug.

**Comment by caweidmann** [comment]
Any update on this when there will be a fix in? We have this issue in our pipeline and I spent the whole day yesterday trying to debug why we are getting inconsistent unit test results with Junit. Alternatively is there a better plugin to use with Jenkins to show coverage?

**Comment by jonlil** [comment]
A workaround for this that we have been using for a long time in many projects now is the following

```bash
sed -i 's/timestamp="[^"]*"/timestamp="'"${TEST_SPLIT_INDEX}"'"/g' \
    ".jenkins/tmp/reports/junit.xml"
```

cc @caweidmann

File: src/main/java/hudson/tasks/junit/TestResult.java (hunk 1 above)
Comment (khneal) [reply src/main/java/hudson/tasks/junit/TestResult.java:307]: > 1. calculate checksums for the report files and compare them
This would break the original use case from years ago, which was separate files per testsuite, then a single file with all testsuites aggregated. The checksum wouldn't match, but the intention was to not add duplicate tests.

> 2. disable this logic with a configuration parameter like
I think more and more people will run into false positives where tests are disappearing and it's difficult to figure out why. If the logic does not work correctly in the default configuration, then maybe it should be removed... which is actually the direction that I've started moving towards. See comment that I will post shortly.

**Comment by khneal** [comment]
Based on all the people that have posted about this issue (in this PR as well as this and other JIRA issues), the deduplication logic is doing far more damage than good. The missing tests have already wasted a lot of person-hours, whereas I would argue the number of hours saved by the cases that have been correctly deduplicated is negligible. If this logic were removed, then the fix for removing duplicate test cases is... just don't include them in the file pattern. Simple. Or delete the aggregated file before the build finishes. Either way, waaay better than randomly missing some tests.

**Comment by khneal** [comment]
I created PR #131 to remove the deduplication logic, so I'm closing this PR.
@jonlil - would you be willing to review that PR?
